jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -1027,54 +1057,68 @@ public void run()
                 Runnable job = null;
-                while (true)
+                exit: while (true)
                 {
                     // If we had a job,
                     if (job != null)
                     {
+                        // signal that we are idle again,
                         idle = true;
-                        // signal that we are idle again
                         if (!addCounts(0, 1))
                             break;
                     }
-                    // else check we are still running
+                    // else check we are still running.
                     else if (_counts.getHi() == Integer.MIN_VALUE)
                     {
                         break;
                     }
 
                     try
                     {
-                        // Look for an immediately available job
-                        job = _jobs.poll();
+                        long idleTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(getIdleTimeout());
+                        job = idleJobPoll(idleTimeoutNanos);
                         if (job == null)
                         {
-                            // No job immediately available maybe we should shrink?
-                            long idleTimeout = getIdleTimeout();
-                            if (idleTimeout > 0 && getThreads() > _minThreads)
+                            // No job available after an idle timeout, maybe we should shrink?
+                            if (idleTimeoutNanos > 0 && getThreads() > getMinThreads())
                             {
-                                long last = _lastShrink.get();
                                 long now = NanoTime.now();
-                                if (NanoTime.millisElapsed(last, now) > idleTimeout && _lastShrink.compareAndSet(last, now))
+                                long shrinkPeriod = idleTimeoutNanos / getIdleTimeoutMaxShrinkCount();
+                                if (LOG.isDebugEnabled())
+                                    LOG.debug("Shrink check, period={}ms {}", TimeUnit.NANOSECONDS.toMillis(shrinkPeriod), QueuedThreadPool.this);
+                                while (true)
                                 {
-                                    if (LOG.isDebugEnabled())
-                                        LOG.debug("shrinking {}", QueuedThreadPool.this);
-                                    break;
+                                    long lastShrink = _lastShrink.get();
+                                    long prevShrink = lastShrink;
+                                    // If the shrink window is too far in the past,
+                                    // advance it to be one idle timeout before now.
+                                    if (NanoTime.elapsed(lastShrink, now) > idleTimeoutNanos)
+                                        prevShrink = now - idleTimeoutNanos;
+
+                                    // Add shrink periods until the window is full.
```

Comment:
I don't like "until" in this comment, implies looping until it is full.
```suggestion
                                    // The next shrink can happen one shrink period after the threshold. 
```

Reply:
But it does not "happen" at that time... `nextShrink` is perhaps badly named (but can't find better) but it is unrelated with _when_ the shrink happens (which is now, not at `nextShrink`).

Reply:
Maybe call it `nextShrinkAfter`.

Eitherway, the "until" text doesn't help an it is better to just say when the next shrink can happen.

```suggestion
                                    // Another shrink is not allowed before one shrink period after the threshhold.
```